### ISSUE TITLE
PERF: System.IO.Win32FileStream.Read allocates buffers unnecessarily

### ISSUE THREAD

**Issue by pharring** [dotnet/corefx#580]
On Win32, if you create a FileStream over a small file (less than 4KB) and try to read the entire contents into a byte array, you'll end up calling this overload:
```c#
public override int Read([In, Out] byte[] array, int offset, int count)
```
On line [731 of Win32FileStream.cs] (
https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs#L731) there is a check to see if `count` is greater than or equal to `_bufferSize` in order to avoid creating the internal 4KB (by default) buffer and, instead, just read the bytes directly into the user's buffer.
However, there's a common case not covered: If the file is shorter than `_bufferSize` and 'count' is exactly the right size, then it ought to be possible to do the same optimization - use the user's buffer and avoid creating the internal buffer.

**Comment by stephentoub** [comment]
@pharring, I think it'd be great to be able to avoid that buffer allocation in such situations.  However, at this point in the code, we don't necessary know the length of the file, right (if the file stream is even seekable)?  Which implies we'd have to do additional I/O (GetFileSizeEx) in order to get it.  Even for the cases where the buffer allocation could be avoided, do we know that would be a net win? And do we know the frequency of such cases such that we wouldn't be overly penalizing the cases where the buffer would still need to be allocated because the file size was bigger?

That said, it seems like there are some related optimizations that could be done.  For example, File.WriteAllBytes isn't currently setting the buffer size to be used by the underlying file stream, but there shouldn't be a need to go through an intermediary buffer in this case, and instead the bufferSize could just be set to the array's Length.

**Comment by pharring** [comment]
I notice that the implementation of [System.IO.File.InternalReadAllBytes] (https://github.com/dotnet/corefx/blob/377dde76c5df105a7bab2c0a3002c0d835ff1b07/src/System.IO.FileSystem/src/System/IO/File.cs#L436) also uses the default buffer size (4KB) which would be needlessly allocated for files smaller than that. Perhaps that could be addressed by specifying a very small buffer (1 byte?)

**Comment by stephentoub** [comment]
Yup, exactly. I'll take a look at a few of these helpers and see what can be done.

**Comment by stephentoub** [comment]
Addressed ReadAllBytes implementation in #590.  Thanks.
